lib/errorUtils.js

```diff
@@ -30,15 +30,25 @@ function causedBy (err, cause) {
   return err
 }
 
+function throwError (err) {
+  throw err
+}
+
 function handlePromiseError (err, reply) {
   if (!(err instanceof Error)) {
     var error = new TypeError(`Attempted to reject a promise with a non-error value from type '${typeof err}'`)
-    reply.res.connection.emit('error', causedBy(error, err))
-    throw error
+    setImmediate(throwError, error)
+    var errWithCause = causedBy(error, err)
+    reply.request.log.error(errWithCause, 'client error')
+    reply.res.connection.destroy(errWithCause)
+    return true
   } else if (isSystem(err)) {
-    reply.res.connection.emit('error', err)
-    throw err
+    setImmediate(throwError, err)
+    reply.request.log.error(err, 'client error')
+    reply.res.connection.destroy(err)
```

Comment:
This is not needed at all.

Reply:
We don't want to close the socket before we exit? The user has to wait for a timeout.

Reply:
Yeah you are right. But `reply.end(new Error('Internal Server Error'))` is better. No?

Reply:
Yeah you are right. But `reply.end(new Error('Internal Server Error'))` is better. No?

Reply:
No, it should signal an unplanned shutdown. It should not be covered by a http status code.

Reply:
@mcollina has also some good reasons for this.

Reply:
I'm not a fan of breaking HTTP protocol digging into the TCP connection.

If there're some reasons to do that, please list here (or send a link) and write the documentation consequentially.

Reply:
If an error is thrown in synchronous code this isn't called, so why call it here?

Reply:
@allevo https://github.com/fastify/fastify/pull/591#pullrequestreview-85897548
If we would return `500` we would express that something happens what the server is able to handle but that's not the case I'm +1 to exit without a status code.

Reply:
@nwoltman we defer the execution to the next `tick` with the help of `setImmediate`. Destroying a socket is a synchronous task.